File: tetris/firmware/mega328/scale.c
// 47 available notes
// song spans 31 notes
// need 5 bits for notes
// lengthmax is 7
// combo set set is 25
#define NNOTES  264
#define NUNIQUE 25
#define NPITCH  24
uint16_t  pitches[NPITCH] = {134, 151, 25, 28, 31, 33, 35, 37, 40, 42, 45, 47, 50, 56, 60, 63, 67, 75, 80, 84, 90, 95, 101, 113};
uint16_t unique[NUNIQUE] = {0x0408, 0x0409, 0x0402, 0x0403, 0x0400, 0x0401, 0x0406, 0x0407, 0x0404, 0x0405, 0x030F, 0x040B, 0x040C, 0x040A, 0x040F, 0x040D, 0x040E, 0x0411, 0x0410, 0x0413, 0x0412, 0x0415, 0x0414, 0x0417, 0x0416};
const uint8_t  notes[NNOTES]   PROGMEM = {12, 15, 12, 14, 12, 18, 12, 17, 12, 20, 12, 17, 12, 18, 12, 14, 12, 24, 12, 22, 12, 20, 12, 17, 12, 20, 12, 17, 12, 18, 12, 14, 7, 1, 7, 11, 7, 12, 7, 15, 7, 16, 7, 15, 7, 12, 7, 11, 7, 17, 7, 18, 7, 16, 7, 15, 7, 16, 7, 15, 7, 12, 7, 23, 17, 7, 17, 23, 17, 7, 17, 6, 19, 15, 19, 6, 19, 15, 19, 19, 14, 12, 14, 19, 14, 12, 14, 21, 17, 14, 17, 21, 17, 14, 17, 21, 18, 15, 18, 21, 18, 15, 18, 24, 20, 18, 20, 24, 20, 18, 20, 5, 17, 14, 17, 5, 17, 14, 17, 4, 22, 20, 24, 4, 24, 20, 24, 5, 24, 17, 24, 5, 24, 17, 24, 4, 15, 21, 15, 4, 15, 21, 23, 14, 2, 18, 2, 17, 2, 15, 2, 14, 2, 18, 2, 14, 2, 0, 2, 7, 2, 12, 2, 17, 2, 12, 2, 14, 2, 18, 2, 17, 2, 0, 2, 12, 2, 9, 2, 14, 2, 1, 2, 15, 2, 1, 2, 7, 3, 13, 3, 19, 3, 7, 3, 18, 3, 11, 3, 14, 3, 11, 3, 1, 8, 12, 8, 21, 8, 1, 8, 5, 8, 12, 8, 18, 8, 12, 8, 11, 9, 15, 9, 24, 9, 11, 9, 12, 9, 15, 9, 5, 9, 15, 9, 12, 7, 24, 7, 21, 9, 21, 9, 24, 7, 24, 7, 18, 0, 11, 0, 12, 7, 14, 7, 21, 9, 21, 9, 24, 7, 24, 0, 10};

File: Projects/tetris/firmware/mega8-16/scale.py
#!/usr/bin/env python
from math import *
"""
'''

E F# G G F# E D# D# E F# B B C# D# E E D  C B B A G F# F# G A B A G F# E E 
0 2  3 3 2 0  11 11 0 2  7 7 9  11 0 0 10 8 7 7 5 3 2 2   3 5 7 5 3 2 0 0
-2 -9 -7 -7 3 -2 


'''
song = map(eval,'''
0 2 3 3 2 0 -1 -1 0 2 -5 -5 -3 -1 0 0 -2 -4 -5 -5 -7 -9 -10 -10 -9 -7 -5 -7 -9 -10 -12 -12 
0 2 3 3 2 0 -1 -1 0 2 -5 -5 -3 -1 0 0 -2 -4 -5 -5 -7 -9 -10 -10 -10 -9 -9
-5 -9 -2 -7 -4 -5 -5 3 -2 0 0 -5 -2 -4 -5 -7 -8 -8 -7 -5 -4 -4 -5  -7 -5 -7 -9 -10 -10
-5 -9 -2 -7 -4 -5 -5 3 -2 0 0 -5 -2 -4 -5 -7 -8 -8 -7 -5 -4 -4 -5  -7 -5 -7 -9 -10 -10
'''.split())
song = map(eval,'''
0 2 3 2 0 -1 0 2 -5 -3 -1 0 -2 -4 -5 -7 -9 -10 -9 -7 -5 -7 -9 -10 -12 
0 2 3 2 0 -1 0 2 -5 -3 -1 0 -2 -4 -5 -7 -9 -10 -9
-5 -9 -2 -7 -4 -5 3 -2 0 -5 -2 -4 -5 -7 -8 -7 -5 -4 -5  -7 -5 -7 -9 -10
-5 -9 -2 -7 -4 -5 3 -2 0 -5 -2 -4 -5 -7 -8 -7 -5 -4 -5  -7 -5 -7 -9 -10
'''.split())
song = map(eval,'''
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19
'''.split())

song = []

for i in range(2,2*6):
	song = song + range(6,6+i) + list(reversed(range(6,6+i-1)))

song = map(eval,'''
0 2 3 2 0 -1 0 2 -5 -3 -1 0 -2 -4 -5 -7 -9 -10 -9 -7 -5 -7 -9 -10 -12 
0 2 3 2 0 -1 0 2 -5 -3 -1 0 -2 -4 -5 -7 -9 -10 -9
-5 -9 -2 -7 -4 -5 3 -2 0 -5 -2 -4 -5 -7 -8 -7 -5 -4 -5  -7 -5 -7 -9 -10
-5 -9 -2 -7 -4 -5 3 -2 0 -5 -2 -4 -5 -7 -8 -7 -5 -4 -5  -7 -5 -7 -9 -10
'''.split())
#song = map(int,'0 2 3 3 2 0 11 11 0 2 7 7 9 11 0 0 10 8 7 7 5 3 2 2 3 5 7 5 3 2 0 0'.split())
"""
song = map(eval,'''24 26 14 29 24 24 29 31 33 35 24 24 31 29 11 26 29 28 17 24 16 26 12 24 7 17 16 14 24 26 14 29 16 12 29 31 33 35 17 16 14 12 31 29 11 26 29 28 16 24 19 26 7 23 36 38 40 36 31 33 35 31 9 12 35 9 31 28 30 14 28 30 31 33 35 19 18 19 11 14 19 23 21 23 16 14 12 19 24 23 24 7 21 23 24 11 28 9 12 11 12 31 4 7 12 7 0'''.split())

execfile('midi.py')

#use = [0,2,4,5,7,9,11] #major
#use = [0,2,3,5,7,8,10] #minor
#use = [0,3,5,6,7,10] #blues
use = range(12)

topf    = 1./18
bottomf = 1./360

k = 0
notes = []
while bottomf*2**(k/12.)<=topf:
	if k%12 in use:
		notes.append( int( 1./(bottomf*2**(k/12.)) ))
	k = k+1

print '// %d available notes'%len(notes)
print '// song spans %d notes'%(max(song)-min(song))

offset = len(notes)-max(song)-1
notes = [notes[i+offset] for i in song]

pitches = list(set(notes))

notebits = ceil(log(len(pitches))/log(2))
lengthbits = 8-notebits
lengthmax = 2**lengthbits-1
print '// need %d bits for notes'%notebits
print '// lengthmax is %d'%lengthmax
notes = ['0x%04X'%(256*d+pitches.index(i)) for d,i in zip(z,notes)]
print '// combo set set is %s'%len(list(set(notes)))

unique = list(set(notes))

notes = [unique.index(i) for i in notes]

print '#define NNOTES  %d'%len(notes)
print '#define NUNIQUE %d'%len(unique)
print '#define NPITCH  %d'%len(pitches)
print 'uint16_t  pitches[NPITCH] = {'+', '.join(map(str,pitches))+'};'
print 'uint16_t unique[NUNIQUE] = {'+', '.join(map(str,unique))+'};'
print 'const uint8_t  notes[NNOTES]   PROGMEM = {'+', '.join(map(str,notes))+'};'



